### ISSUE TITLE
Generation fails

### ISSUE THREAD

**Issue by fonta** [aequasi/discord-service-definition-generator#1]
It looks like the generator fails. The json returned on the page has lot and lots of emptiness.
Possibly an update is needed.

![Example of result](https://i.imgur.com/4IHswTO.png)

**Comment by warlof** [comment]
Apparently, Discord alter the way they're rendering the Documentation portal.
All the HTML rendering have been moved to the browser.

Therefore, the library is only showing the basic skeleton :

```
<p>&lt;!DOCTYPE html&gt;</p>
<html>
  <head>
    <meta charset="utf-8">
    <meta content="width=device-width, initial-scale=1.0, maximum-scale=1, user-scalable=no" name="viewport">

    <meta property="og:type" content="website">
    <meta property="og:site_name" content="Discord Developer Portal">
    <meta property="og:title" content="Discord Developer Portal &#x2014; API Docs for Bots and Developers">
    <meta property="og:description" content="Integrate your service with Discord &#x2014; whether it&apos;s a bot or a game or your whatever you wildest imagination can come up with.">

    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:site" content="@discordapp">
    <meta name="twitter:creator" content="@discordapp">

    <link rel="chrome-webstore-item" href="https://chrome.google.com/webstore/detail/lcbhdgefieegnkbopmgklhlpjjdgmbog">
<link rel="stylesheet" href="/assets/0.f0ee59761b853f90a44e.css" integrity="sha256-4JxTxzIDOkOY7Mqw4rYtaanKKuMYQN9C673+mRdkXVA= sha512-GHdW1CenOLNvhadSlJjb+O0YvnzzlJtvD0jSKWddE4FWtNUJZ+RXG2jlDcEL/bLvxBDxJe+zX20vaPHj11IQyg=="><link rel="icon" href="/assets/07dca80a102d4149e9736d4b162cff6f.ico">    <title>Discord Developer Portal</title>
  </head>

  <body>
    <div id="app-mount"></div><script>window.GLOBAL_ENV = {
      API_ENDPOINT: &#39;//discordapp.com/api&#39;,
      WEBAPP_ENDPOINT: &#39;//discordapp.com&#39;,
      CDN_HOST: &#39;cdn.discordapp.com&#39;,
      ASSET_ENDPOINT: &#39;<a href="http://localhost:3333&#39;">http://localhost:3333&#39;</a>,
      WIDGET_ENDPOINT: &#39;//localhost:3335&#39;,
      INVITE_HOST: &#39;undefined&#39;,
      GIFT_CODE_HOST: &#39;undefined&#39;,
      MARKETING_ENDPOINT: &#39;//discordapp.com&#39;,
      RELEASE_CHANNEL: &#39;stable&#39;,
      BRAINTREE_KEY: &#39;production_5st77rrc_49pp2rp4phym7387&#39;,
      STRIPE_KEY: &#39;pk_live_CUQtlpQUF0vufWpnpUmQvcdi&#39;,
    };</script><script src="/assets/51bcd0bbf16930d50d61.js" integrity="sha256-ccm0GEZMvo2LUbnT3KtWGTgTdGrv5oyD/OGKNbztS5Y= sha512-+gwoGfmDyPcazsD/2xcuYzxH/0J7boIufwxq29YYAhtlu8aIgMIz+U5jBtrlN1C872MJCtxxZVYg7cZnvi/z0Q=="></script><script src="/assets/0c46e8c01518144dfc3c.js" integrity="sha256-OouywhpLwSAo6ieIJXwJUQ8S2dyCwK/RuV4iSN/83Fs= sha512-/iXr65j5vjKoI8Kviu5AahbtYdXHigpcSnPiVj008ZNDLCeaDLHMPIwI4FmGWWwraNfMZ5bszleKkSUuQUHFeQ=="></script><script src="/assets/8d5c9e907ed0ecb514bb.js" integrity="sha256-7cIGWzaw9TYAyExJxlwbr4yeBc9ioG87SozZzfbg2fo= sha512-4t5OUO5AA0Bu/xWP+FT+MdMFw4b+LTrqQ9pB2dhl0TG3ymkT3jYZ/+ManCYPE99lySwSmst4BobFBtMn18y+qA=="></script>  </body>
</html>
```

**Comment by fonta** [comment]
It indeed looks like that.
Hopefully @aequasi can update the script.

**Comment by fonta** [comment]
I dug a little bit into the code.
It seems that discord changed the way the docs are loaded.
Nowadays it loads the contents using JS into basic skeleton.
Cheerio simply parses the html without waiting for the contents to be loaded.
That's not a bug, because cheerio was simply not designed to do that.

After manually saving all the different pages into the cache directory, the parser did generate a "complete" new service definition. However it seems that more has changed because with a simple compare it shows that it doesn't generate response types, and complete sections are missing.

![image](https://user-images.githubusercontent.com/6190791/57197466-deb18000-6f67-11e9-99aa-1b007ac6e633.png)

For example, it skips the whole channel page.

Next to that it changes "ISO8601 timestamp" into integer, and bool is put out as boolean.

In short:
- it doesn't wait for a full page load
- it doesn't generate responseTypes
- it skips sections
- it change timestamp to integer
- bool is changed in boolean

**Comment by fonta** [comment]
Found out that it gives errors on the invite and channel pages:

![image](https://user-images.githubusercontent.com/6190791/57197926-1b33aa80-6f6d-11e9-8e4e-eac7ec2becde.png)

In the invite.html, the following table generates the error:

![image](https://user-images.githubusercontent.com/6190791/57198144-c0e81900-6f6f-11e9-91e3-7e3e84fc3cb3.png)

So after removing the following html from the file, it has no error anymore:
```html
<h5 id="target-user-type">
    Target User Type
    <a class="anchor-3Z-8Bb hyperlink" href="https://discordapp.com/developers/docs/resources/invite#target-user-type">
        <div name="target-user-type"></div>
    </a>
</h5>
<table>
    <thead>
        <tr>
            <th scope="col">Type</th>
            <th scope="col">Value</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>STREAM</td>
            <td>1</td>
        </tr>
    </tbody>
</table>
```

---

## edit 2

Thought that shouldn't be the solution, so I added a check if type = undefined to return.

```Javascript
if (type == undefined) {
    return;
}
```

![image](https://user-images.githubusercontent.com/6190791/57198407-9b104380-6f72-11e9-823b-e77e34a877ce.png)

The now generated json does contain the channel and invite sections :)

---

## edit 3
Found out why the responsetypes were not generated.
Cheerio was unable to find the a elements in the match.
This was because the regex cut of after "discordapp."

![image](https://user-images.githubusercontent.com/6190791/57199272-61910580-6f7d-11e9-951d-0589bb3d4c8b.png)

So with a little adjustment, it now gets the whole link object:

![image](https://user-images.githubusercontent.com/6190791/57199226-fb0be780-6f7c-11e9-84af-f97a99ed13ce.png)

So in the code, I added a uriRegex to get the url html code, and used that regex continuing forward.
```Javascript
let uriRegex = /(Return[^\.]+.+>)/g;
let uriMatch = uriRegex.exec(desc.html());
if (uriMatch !== null) {
    cheerio(`<div>${uriMatch[0]}</div>`).find('a').each((i, e) => {
        const href         = $(e).attr('href').split("#");
        let returnResource = href[0].split('/').reverse()[0];
        let object         = href[1];
        if (object.indexOf('-object') === -1) {
            return;
        }
        
        // @todo Remove once https://github.com/discordapp/discord-api-docs/pull/501 is merged
        if (key === 'listGuildMembers') {
            returnResource = "guild";
            object         = "guild-member";
        }
        
        const array = listRegex.test(uriMatch[0]);
        let type    =
                    (array ? 'Array<' : '') +
                    (returnResource ? (returnResource + "/") : "") +
                    object.replace('-object', '').replace('DOCS_', '') +
                    (array ? '>' : '');
        
        responseTypes.push({
            name: $(e).text(),
            type
        });
    });
}
```

![image](https://user-images.githubusercontent.com/6190791/57199486-a7e76400-6f7f-11e9-9a91-a8c975846eeb.png)

The result is at this moment a generated service definition with all sections and the responstypes includes.

---

## edit 4
I've change line 312 so that ISO8601 timestamps are returned as ISO8601 timestamp and not as integers which php can't understand as timestamps.

![image](https://user-images.githubusercontent.com/6190791/57199593-3c05fb00-6f81-11e9-9984-064644e4fc00.png)

The only difference left now is that bool is now boolean. But that doesn't seem to be an issue.
And also a solution will need to be found for the script only downloading the page framework.

**Comment by fonta** [comment]
I've created a pull request for the small changes I made described in the post above.
With manually saving the pages to the cache folder, it now creates a usable service definition.

**Comment by fonta** [comment]
I was able to get the downloading of the complete page fixed by using puppeteer. 
But because the whole script uses Cheerio, it would be some kind of a nasty workaround.
It's also a bit slower than cheerio it seems.
Maybe domjs is a solution.

**Comment by warlof** [comment]
Q@f0nt4 I tried domjs without success. Maybe I've miss something while using it.
What about using puppeteer to only seed the cache?

An alternative would be to use marked in order to generate the docs based on the repository. 
There is a way to customize the way markdown is rendered, so we could reduce the impact on existing code.

**Comment by fonta** [comment]
@warlof That's indeed what I did, added puppeteer with `npm i puppeteer` and added some code to download the page to the html file. When that's finished I used Cheerio again to retrieve the file and return in to the script.
domJS was also unsuccessful for me, but probably because of my lack of knowledge of the package.

Don't know about Marked, but can imagine that it's also a good solution as the Github raw content doesn't have to deal with dynamic loaded content.

Will see if I have some time this evening.

**Comment by fonta** [comment]
This is a working example using Puppeteer:

After `npm i puppeteer --save' replace this function in parser.js.
```Javascript
async getResource(resource) {
    let type = this.topics.indexOf(resource) === -1 ? 'resources' : 'topics';

    let dir = path.join(__dirname, '..', 'cache', type);
    let file = path.join(dir, resource + ".html");
    mkdirp.sync(dir);

    try {
        return cheerio.load(fs.readFileSync(file));
    } catch (error) {
        const url = `${baseUrl}/${type}/${resource}`;

        const puppeteer = require('puppeteer');
        const browser = await puppeteer.launch();

        const page = await browser.newPage();
        await page.goto(url);
        await page.waitForSelector('[class^=contentWrapperInner]') 

        const content = await page.content()
        await browser.close();

        //fs.writeFileSync(file, content);

        return cheerio.load(content);
    }
}
```

It is pretty slow though, and saving the content to files is commented out.

**Comment by fonta** [comment]
@warlof Pull Requests have been done. Just waiting for the last one to get merged.
Also @aequasi will have to add the Puppeteer buildpack to the app on Heroku.
You can see it to work here: https://testdiscorddefgen.herokuapp.com/
But I'll remove that app after @aequasi merged the last [PR](https://github.com/aequasi/discord-service-definition-generator/pull/5)

**Comment by fonta** [comment]
There you go! https://discord-service-definition.herokuapp.com/ is back in business 👯
